File: frontend/src/app/service/horse.service.ts
import {HttpClient, HttpParams} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {Observable} from 'rxjs';
import {environment} from 'src/environments/environment';
import {Horse, HorseSearch} from '../dto/horse';
import {Sex} from '../dto/sex';

const baseUri = environment.backendUrl + '/horses';

@Injectable({
  providedIn: 'root'
})
export class HorseService {

  constructor(
    private http: HttpClient,
  ) {
  }

  /**
   * Get all horses stored in the system
   *
   * @return observable list of found horses.
   */
  getHorsesMatchingParams(horse: HorseSearch): Observable<Horse[]> {
    let params = new HttpParams();
    if (horse.name !== null && horse.name !== undefined) {
      params = params.append('name', horse.name);
    }
    if (horse.description !== null && horse.description !== undefined && horse.description !== '') {
      params = params.append('description', horse.description);
    }
    if (horse.sex !== null && horse.sex !== undefined) {
      params = params.append('sex', horse.sex);
    }
    if (horse.bornBefore !== null && horse.bornBefore !== undefined) {
      params = params.append('bornBefore', horse.bornBefore.toString());
    }
    if (horse.owner !== null && horse.owner !== undefined && horse.owner !== '') {
      params = params.append('ownerName', horse.owner);
    }
    return this.http.get<Horse[]>(baseUri, {params});
  }


  /**
   * Create a new horse in the system.
   *
   * @param horse the data for the horse that should be created
   * @return an Observable for the created horse
   */
  create(horse: Horse): Observable<Horse> {
    return this.http.post<Horse>(
      baseUri,
      horse
    );
  }

  /**
   * Get Horse by ID
   *
   * @param id of the horse to retrieve
   * @return an Observable of the horse with the given id
   */
  getById(id: number): Observable<Horse> {
    return this.http.get<Horse>(baseUri + '/' + id);
  }

  /**
   * Edit an existing Horse
   *
   * @param horse the new data for the horse to update
   * @return an Observable of the edited horse
   */
  edit(horse: Horse): Observable<Horse> {
    return this.http.put<Horse>(baseUri + '/' + horse.id, horse);
  }

  /**
   * Delete an existing Horse
   *
   * @param id of the horse to be deleted
   * @return an Observable of the deleted horse
   */
  delete(id: number): Observable<Horse> {
    return this.http.delete<Horse>(baseUri + '/' + id);
  }

  /**
   * Searches for Name suggestions for possible Mothers for a horse
   *
   * @param dateOfBirth birthdate of the horse for which possible mothers should be searched for
   * @param name name or substring of the name of possible mothers
   * @param limitTo limits the result list to the specified number
   * @return an Observable of the list of the matching Mothers
   */
  public searchByNameMother(dateOfBirth: Date | undefined, name: string, limitTo: number): Observable<Horse[]> {
    const params = new HttpParams()
      .set('name', name)
      .set('sex', Sex.female)
      .set('bornBefore', dateOfBirth !== undefined ? dateOfBirth.toString() : '')
      .set('limit', limitTo);
    return this.http.get<Horse[]>(baseUri, {params});
  }

  /**
   * Searches for Name suggestions for possible Fathers for a horse
   *
   * @param dateOfBirth birthdate of the horse for which possible fathers should be searched for
   * @param name name or substring of the name of possible fathers
   * @param limitTo limits the result list to the specified number
   * @return an Observable of the list of the matching Fathers
   */
  public searchByNameFather(dateOfBirth: Date | undefined, name: string, limitTo: number): Observable<Horse[]> {
    const params = new HttpParams()
      .set('name', name)
      .set('bornBefore', dateOfBirth !== undefined ? dateOfBirth.toString() : '')
      .set('sex', Sex.male)
      .set('limit', limitTo);
    return this.http.get<Horse[]>(baseUri, {params});
  }

  /**
   * Retrieves the familyTree for a given Horse as the root horse, meaning all ancestors of the root horse
   * up to a given generation
   *
   * @param id of the root horse
   * @param generations number of generations which should be returned
   * @return an Observable of the FamilyTree with the horse with the given id beeing the root of the tree
   */
  public getFamilyTree(id: number, generations: number): Observable<Horse> {
    const params = new HttpParams().set('generations', generations);
    return this.http.get<Horse>(baseUri + '/' + id + '/familytree', {params});
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/assignment/individual/service/HorseService.java
package at.ac.tuwien.sepm.assignment.individual.service;

import at.ac.tuwien.sepm.assignment.individual.dto.HorseCreateDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseDetailDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseFamilyTreeDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseListDto;
import at.ac.tuwien.sepm.assignment.individual.dto.HorseSearchDto;
import at.ac.tuwien.sepm.assignment.individual.exception.ConflictException;
import at.ac.tuwien.sepm.assignment.individual.exception.NotFoundException;
import at.ac.tuwien.sepm.assignment.individual.exception.ValidationException;

import java.util.stream.Stream;

/**
 * Service for working with horses.
 */
public interface HorseService {
  /**
   * Lists all horses stored in the system.
   *
   * @return list of all stored horses
   */
  Stream<HorseListDto> allHorses();


  /**
   * Updates the horse with the ID given in {@code horse}
   * with the data given in {@code horse}
   * in the persistent data store.
   *
   * @param horse the horse to update
   * @return the updated horse
   * @throws NotFoundException   if the horse with given ID does not exist in the persistent data store
   * @throws ValidationException if the update data given for the horse is in itself incorrect (description too long, no name, …)
   * @throws ConflictException   if the update data given for the horse is in conflict the data currently in the system (child older than parent, …)
   */
  HorseDetailDto update(HorseDetailDto horse) throws NotFoundException, ValidationException, ConflictException;


  /**
   * Get the horse with given ID, with more detail information.
   * This includes the owner of the horse, and its parents.
   * The parents of the parents are not included.
   *
   * @param id the ID of the horse to get
   * @return the horse with ID {@code id}
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  HorseDetailDto getById(long id) throws NotFoundException;

  /**
   * Create a new Horse with Data given in {@code horse}
   * stores the horse in the persistent data store and
   * generates a unique ID.
   *
   * @param horse the horse to create
   * @return Created Horse with generated ID
   * @throws ValidationException if the data given for the horse is in itself incorrect(description too long, no name, ...)
   * @throws ConflictException   if the data given for the horse is in conflict with the data currently in the system (child older than parent, ...)
   */
  HorseDetailDto create(HorseCreateDto horse) throws ValidationException, ConflictException;

  /**
   * Deletes a Horse with the id given {@code id}
   * deletes every relationship including this horse(parent-child
   * , child-parent, owner-horse)
   *
   * @param id the ID of the horse to delete
   * @throws NotFoundException if the horse with the given ID does not exist in the persistent data store
   */
  void delete(long id) throws NotFoundException;

  /**
   * Search for Horses matching the criteria in {@code searchParameters}.
   * <p>
   * A Horse is considered matched, if it matches all the criteria given in {@code searchParams}
   * Strings ,name/description/owner , are matched if it contains {@code searchParams.name}/{@code searchParams.description}
   * /{@code searchParams.owner} as a substring.
   * The birthdate is matched if it is before {@code searchParams.bornBefore}
   * Sex needs to be an exact match, male or female
   * If searchParams.limit is not null the returned stream of horses never
   * contains more than {@code searchParams.limit} elements,
   * even if there would be more matches in the persistent data store.
   * If searchParams.limit is null every matching element is returned in the stream
   * </p>
   *
   * @param searchParams object containing the search parameters to match
   * @return a Stream containing horses matching the criteria in {@code searchParams}
   */
  Stream<HorseListDto> searchForHorses(HorseSearchDto searchParams);

  /**
   * get the Ancestors of the specified(by id) horse as a {@link HorseFamilyTreeDto}
   * up to (including) the given generation
   *
   * @param id          id of the root horse of the family Tree
   * @param generations number of generations which are extracted
   * @return a {@link HorseFamilyTreeDto} representing the FamilyTree of the horse
   * @throws NotFoundException if Horse with id is not found in system
   */
  HorseFamilyTreeDto getFamilyTree(Long id, Long generations) throws NotFoundException;
}
